pkg/kubevirt/data_volume.go

```diff
@@ -0,0 +1,118 @@
+	client ClientFunc
+}
+
+// NewDefaultDataVolumeManager creates a new default manager with a k8s controller-runtime based on the sent kubecomfig.
+func NewDefaultDataVolumeManager(client ClientFunc) (DataVolumeManager, error) {
+	return &defaultDataVolumeManager{
+		client: client,
+	}, nil
+}
+
+// CreateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+func (d *defaultDataVolumeManager) CreateVolume(ctx context.Context, secret *v1.Secret, name, namespace string, dvSpec cdi.DataVolumeSpec) error {
+	dataVolume := &cdi.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: dvSpec,
+	}
+
+	c, err := d.client(secret)
+	if err != nil {
+		return fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+	return c.Create(ctx, dataVolume)
+}
+
+// GetVolume fetches the specified volume by the passed name and namespace and return DataVolumeNotFoundError error in case of
+// not found object error.
+func (d *defaultDataVolumeManager) GetVolume(ctx context.Context, secret *v1.Secret, name, namespace string) (*cdi.DataVolume, error) {
+	c, err := d.client(secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+
+	dataVolume := &cdi.DataVolume{}
+	if err := c.Get(ctx, client.ObjectKey{Name: name, Namespace: namespace}, dataVolume); err != nil {
+		if kerrors.IsNotFound(err) {
+			return nil, &clouderrors.DataVolumeNotFoundError{
+				Name: name,
+			}
+		}
+
+		return nil, fmt.Errorf("failed to get DataVolume: %s: %v", name, err)
+	}
+
+	return dataVolume, nil
+}
+
+// ListDataVolumes lists all the Data Volumes which exists in the passed namespace.
+func (d *defaultDataVolumeManager) ListDataVolumes(ctx context.Context, secret *v1.Secret, namespace string) (*cdi.DataVolumeList, error) {
+	c, err := d.client(secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+
+	dvList := cdi.DataVolumeList{}
+	if err := c.List(ctx, &dvList); err != nil {
+		return nil, fmt.Errorf("failed to list DataVolumes in namespace %s: %v", namespace, err)
+	}
+
+	if len(dvList.Items) == 0 {
+		klog.V(2).Infof("namespace %s has no data volumes", namespace)
+		return nil, nil
+	}
+
+	return &dvList, nil
+}
+
+// DeleteVolume delete the DataVolume based on the passed name and namespace.
+func (d *defaultDataVolumeManager) DeleteVolume(ctx context.Context, secret *v1.Secret, name, namespace string) error {
+	c, err := d.client(secret)
+	if err != nil {
+		return fmt.Errorf("failed to create kubevirt client: %v", err)
+	}
+
+	dv, err := d.GetVolume(ctx, secret, name, namespace)
+	if err != nil {
+		if clouderrors.IsDataVolumeNotFoundError(err) {
+			klog.V(2).Infof("skip DataVolume deletion, DataVolume %s is not found", name)
+			return nil
+		}
+
+		return fmt.Errorf("failed to fetch data volume prior creation: %v", err)
+	}
+
+	return c.Delete(ctx, dv)
```

Comment:
Please handle `NotFound` error during deletion. The fact that the previous call to `GetVolume` returned something doesn't matter, it's a distributed system and therefore the object may disappear between the 2 calls. In fact, I would skip the call to `GetVolume` entirely and only handle `NotFound` during the deletion.

Reply:
And how would you delete the volume with the object itself?

Reply:
You only need to initialize the name and namespace of an object in order to delete it, e.g.:
```go
                dv := &cdi.DataVolume{
			ObjectMeta: metav1.ObjectMeta{
				Name:      name,
				Namespace: namespace,
			},
		}
		if err := client.IgnoreNotFound(a.client.Delete(ctx, dv)); err != nil {
			return errors.Wrap(err, "could not delete datavolume")
		}
```